### PR TITLE
Remove kernelMapping setting and handle kernelspecs asynchronous

### PR DESCRIPTION
**Pull request by lgeiger**
The first commit removes the `kernelMapping` setting. Instead the user is asked to chose a kernel to start.

The second commit introduces callbacks for all kernel specs handling. In order to update the kernel specs if necessary.

It probably makes sense to review each commit on its own.

### CODE REVIEW

File: lib/kernel-manager.coffee [hunk 1]
```diff
@@ -111,26 +112,23 @@ class KernelManager
...
+            @kernelPicker = new KernelPicker ->
+                return kernelSpecs
+            @kernelPicker.onConfirmed = ({kernelSpec}) ->
+                callback kernelSpec
+            @kernelPicker.toggle()
 
 
     kernelSpecProvidesLanguage: (kernelSpec, language) ->
         kernelLanguage = kernelSpec.language
-        mappedLanguage = Config.getJson('languageMappings')[kernelLanguage]
```
Comment (n-riesco) [inline lib/kernel-manager.coffee]: If `languageMappings` is removed, then users won't be able to use kernels with non-standard language names.

File: lib/main.coffee [hunk 2]
```diff
@@ -105,17 +105,17 @@ module.exports = Hydrogen =
             kernel.interrupt()
 
         else if command is 'restart-kernel'
+            kernelSpec = kernel.kernelSpec
```
Comment (n-riesco) [inline lib/main.coffee]: Neat!

File: lib/kernel-manager.coffee [hunk 3]
```diff
@@ -98,33 +90,39 @@ class KernelManager
...
-            callback kernelSpecs[0]
-        else
-            @kernelPicker = new KernelPicker ->
-                return kernelSpecs
-            @kernelPicker.onConfirmed = ({kernelSpec}) ->
-                callback kernelSpec
-            @kernelPicker.toggle()
+        @getAllKernelSpecsFor language, (kernelSpecs) ->
+            if kernelSpecs.length <= 1
+                callback kernelSpecs[0]
```
Comment (n-riesco) [inline lib/kernel-manager.coffee]: `callback kernelSpecs[0] or []`

File: lib/kernel-manager.coffee [hunk 4]
```diff
@@ -111,26 +112,23 @@ class KernelManager
...
         unless language?
             return null
 
-        kernelMapping = Config.getJson('kernelMappings')?[language]
-        if kernelMapping?
-            kernelSpecs = @getAllKernelSpecs().filter (spec) ->
-                return spec.display_name is kernelMapping
+        kernelSpecs = @getAllKernelSpecsFor language
+        if kernelSpecs.length <= 1
+            callback kernelSpecs[0]
```
Comment (n-riesco) [inline lib/kernel-manager.coffee]: `callback kernelSpecs[0] or []`

File: lib/kernel-manager.coffee (hunk 1 above)
Comment (lgeiger) [inline lib/kernel-manager.coffee]: Ooops, I removed it by accident. Thanks for noticing!

File: lib/main.coffee [hunk 5]
```diff
@@ -101,18 +101,25 @@ module.exports = Hydrogen =
         unless kernel
             kernel = @kernelManager.getRunningKernelFor language
 
+        message = "No running kernel for language `#{language}` found"
```
Comment (n-riesco) [inline lib/main.coffee]: I think this kind of early optimisation makes the code less readable (please, read this comment just as an opinion).

File: lib/main.coffee [hunk 6]
```diff
@@ -101,18 +101,25 @@ module.exports = Hydrogen =
...
+                atom.notifications.addError message
+                return
             kernelSpec = kernel.kernelSpec
             @kernelManager.destroyRunningKernel kernel
             @clearResultBubbles()
             @kernelManager.startKernel kernelSpec, grammar, =>
                 @setStatusBarElement()
 
         else if command is 'switch-kernel'
-            kernel = @kernelManager.getRunningKernelFor language
```
Comment (n-riesco) [inline lib/main.coffee]: I think this line is potentially necessary, since the user can pass any kernel.

File: lib/main.coffee (hunk 5 above)
Comment (lgeiger) [inline lib/main.coffee]: I think I'll rename it to `errorMessage` which should make it clearer.

File: lib/main.coffee (hunk 6 above)
Comment (lgeiger) [inline lib/main.coffee]: `handleKernelCommands` is only used with the `command` and `kernelSpec` arguments.
@n-riesco Should I simplify the function and remove the unused arguments?

Comment (lgeiger) [inline lib/main.coffee]: I'm wrong `SignalListView` uses all the arguments though I think this can be simplified in a future PR.

**Comment by lgeiger** [comment]
@n-riesco Thanks for the feedback.

**Comment by rgbkrk** [comment]
Thank you both!

**Pull request merged**